lib/routes/token.js

```diff
@@ -11,21 +12,107 @@ const config = require('../config');
 const db = require('../db');
 const encrypt = require('../encrypt');
 const logger = require('../logging')('routes.token');
+const P = require('../promise');
+const Scope = require('../scope');
 const validators = require('../validators');
 
+const MAX_TTL = config.get('expiration.accessTokenS');
+const AUTHORIZATION_CODE = 'authorization_code';
+const REFRESH_TOKEN = 'refresh_token';
 
-function generateToken(code) {
-  return [code.authAt, db.removeCode(code.code), db.generateToken(code)];
+function confirmClient(id, secret) {
+  return db.getClient(buf(id)).then(function(client) {
+    if (!client) {
+      logger.debug('client.notFound', { id: id });
+      throw AppError.unknownClient(id);
+    }
+
+    var submitted = hex(encrypt.hash(buf(secret)));
+    var stored = hex(client.secret);
+    if (submitted !== stored) {
+      logger.info('client.mismatchSecret', { client: id });
+      logger.verbose('client.mismatchSecret.details', {
+        submitted: submitted,
+        db: stored
+      });
+      throw AppError.incorrectSecret(id);
+    }
+
+    return client;
+  });
 }
 
-function toToken(authAt, _, token) {
-  /*jshint camelcase: false*/
-  return {
-    access_token: token.token.toString('hex'),
-    token_type: token.type,
-    scope: token.scope.join(' '),
-    auth_at: authAt
-  };
+function confirmCode(id, code) {
+  return db.getCode(buf(code)).then(function(codeObj) {
+    if (!codeObj) {
+      logger.debug('code.notFound', { code: code });
+      throw AppError.unknownCode(code);
+    } else if (hex(codeObj.clientId) !== hex(id)) {
+      logger.debug('code.mismatch', {
+        client: hex(id),
+        code: hex(codeObj.clientId)
+      });
+      throw AppError.mismatchCode(code, id);
+    } else {
+      // + because loldatemath. without it, it does string concat
+      var expiresAt = +codeObj.createdAt + config.get('expiration.codeMs');
+      if (Date.now() > expiresAt) {
+        logger.debug('code.expired', { code: code });
+        throw AppError.expiredCode(code, expiresAt);
+      }
+    }
+    return codeObj;
+  });
+}
+
+function confirmToken(params) {
+  return db.getRefreshToken(encrypt.hash(params.refresh_token))
+  .then(function(tokObj) {
+    if (!tokObj) {
+      logger.debug('refresh_token.notFound', params.refresh_token);
+      throw AppError.invalidToken();
+    } else if (hex(tokObj.clientId) !== hex(params.client_id)) {
+      logger.debug('refresh_token.mismatch', {
+        client: params.client_id,
+        code: tokObj.clientId
+      });
+      throw AppError.invalidToken();
+    } else if (!Scope(tokObj.scope).has(params.scope)) {
+      logger.debug('refresh_token.invalidScopes', {
+        allowed: tokObj.scope,
+        requested: params.scope
+      });
+      throw AppError.invalidScopes();
+    }
+    tokObj.scope = params.scope;
+    return tokObj;
+  });
+}
+
+function generateToken(options) {
```

Comment:
suggest `generateTokens` plural to make it clear that multiple tokens might be returned